Network altair gossip and discovery

pawanjay176 opened the pull request:
## Proposed Changes

Implements gossip and discovery changes for altair hard fork. Builds on `network-altair-rpc` branch.

pawanjay176 committed:
Start adding sync committee attnets

pawanjay176 committed:
Add chainspec and genesis validators root to eth2_libp2p

pawanjay176 committed:
Add chainspec and genesis validators root to eth2_libp2p

pawanjay176 committed:
Add chainspec and genesis validators root to eth2_libp2p

pawanjay176 committed:
Add chainspec and genesis validators root to eth2_libp2p

pawanjay176 committed:
Add altair gossip types

pawanjay176 committed:
Subscribe to required gossip topics

pawanjay176 committed:
Unsubscribe from pre fork topics

pawanjay176 committed:
Compiles

pawanjay176 committed:
Add Subnet enum to include SyncCommittee subnet_ids

pawanjay176 committed:
update_enr_bitfield takes a Subnet instead of SubnetId

pawanjay176 committed:
Fix network service

pawanjay176 committed:
Add MetaData V2

pawanjay176 committed:
Don't read context bytes for MetaData V2

pawanjay176 committed:
Add sync committee topics to whitelist

pawanjay176 commented:
@AgeManning would appreciate an initial pass over this guy as well. 
It seems to be running fine on prater for now. Also tested with all local testnets where all nodes run rpc v2.

This does not contain the changes for a sync committee subnet service which makes subscriptions for sync committee subnets yet.

pawanjay176 committed:
Network altair rpc (#2301)

* Add v2 messages to rpc decoder

* Ugly hack

* Pass chainspec and genesis_root to Rpc

* Add context bytes while encoding

* Add a ForkContext struct

* Pass ForkContext to rpc

* crate compiles

* Extract ForkContext into separate file; add a current_fork field

* Fix encoding/decoding

* Fix tests

* Remove fork_schedule from rebase

* Fix ForkContext

* Fix tests

* Remove fork_schedule again

* Add altair empty and full block limits

* Fix panic in snappy decoding

* Fix limits

* Move wrapping of RPCRequests to handler

* RpcRequestContainer only used in OutboundUpgrade

* Add altair blocks in rpc end to end tests

* same rpc limits for V1 and V2

* V2 response decoding happens only for valid protocols

* Add snappy response decoding tests

* Add more snappy tests

* Minor fixes

* Appease clippy

* to_context_bytes returns an Option

* Add padding snappy message test for v2

* Minor fixes; remove accidentally added file

* lint

pawanjay176 committed:
Network altair rpc (#2301)

* Add v2 messages to rpc decoder

* Ugly hack

* Pass chainspec and genesis_root to Rpc

* Add context bytes while encoding

* Add a ForkContext struct

* Pass ForkContext to rpc

* crate compiles

* Extract ForkContext into separate file; add a current_fork field

* Fix encoding/decoding

* Fix tests

* Remove fork_schedule from rebase

* Fix ForkContext

* Fix tests

* Remove fork_schedule again

* Add altair empty and full block limits

* Fix panic in snappy decoding

* Fix limits

* Move wrapping of RPCRequests to handler

* RpcRequestContainer only used in OutboundUpgrade

* Add altair blocks in rpc end to end tests

* same rpc limits for V1 and V2

* V2 response decoding happens only for valid protocols

* Add snappy response decoding tests

* Add more snappy tests

* Minor fixes

* Appease clippy

* to_context_bytes returns an Option

* Add padding snappy message test for v2

* Minor fixes; remove accidentally added file

* lint

pawanjay176 committed:
Network altair rpc (#2301)

* Add v2 messages to rpc decoder

* Ugly hack

* Pass chainspec and genesis_root to Rpc

* Add context bytes while encoding

* Add a ForkContext struct

* Pass ForkContext to rpc

* crate compiles

* Extract ForkContext into separate file; add a current_fork field

* Fix encoding/decoding

* Fix tests

* Remove fork_schedule from rebase

* Fix ForkContext

* Fix tests

* Remove fork_schedule again

* Add altair empty and full block limits

* Fix panic in snappy decoding

* Fix limits

* Move wrapping of RPCRequests to handler

* RpcRequestContainer only used in OutboundUpgrade

* Add altair blocks in rpc end to end tests

* same rpc limits for V1 and V2

* V2 response decoding happens only for valid protocols

* Add snappy response decoding tests

* Add more snappy tests

* Minor fixes

* Appease clippy

* to_context_bytes returns an Option

* Add padding snappy message test for v2

* Minor fixes; remove accidentally added file

* lint

pawanjay176 committed:
Fix metadata encoding; fmt

pawanjay176 committed:
Fix next_fork methods in chainspec

pawanjay176 committed:
Minor fixes

pawanjay176 committed:
Fix next_fork_update in network service; add next_unsubscribe delay

pawanjay176 committed:
Accept peers which haven't updated next_fork_version/epoch

pawanjay176 committed:
Add sync committee topic tests

pawanjay176 committed:
Simplify required_gossip_digests; improve logging around fork boundary

pawanjay176 committed:
Fix tests

pawanjay176 committed:
Minor refactorings

pawanjay176 commented on beacon_node/eth2_libp2p/src/discovery/mod.rs:
```diff
@@ -727,7 +768,11 @@ impl<TSpec: EthSpec> Discovery<TSpec> {
         // predicate for finding nodes with a matching fork and valid tcp port
         let eth2_fork_predicate = move |enr: &Enr| {
-            enr.eth2() == Ok(enr_fork_id.clone()) && (enr.tcp().is_some() || enr.tcp6().is_some())
+            // `next_fork_epoch` and `next_fork_version` can be different so that
```
@AgeManning this might cause compatibility issues with lighthouse nodes that are not altair enabled. 
A non altair enabled node will reject an altair node during discovery because it requires the entire enr fork id to be the same. However, nodes with different `next_fork_epoch` and `next_fork_version`  in their enr fork id should still be accepted according to the spec.

AgeManning replied:
Can we just match on either current fork digest or next fork version? 
I think its fine to discover peers who say they are on the next fork digest. After we dial we'll kick them if they are not compatible.

pawanjay176 replied:
Didn't get you. 
I'm saying if node A (old networking) runs a discovery query and finds node B (altair networking), then it will reject node B in the discovery step because node A is running predicate `enr_fork_digest_a == enr_fork_digest_b` which is `false`. So old nodes will not be able to discover altair nodes (other way works fine). Not sure if I'm missing something here.

pawanjay176 replied:
Tested this in a local testnet and it is an issue. The non-altair node rejects the altair nodes. Made a PR to fix the issue here https://github.com/sigp/lighthouse/pull/2433

pawanjay176 commented:
@AgeManning this is passing all tests and ready for review. 
I have tested it out on prater for backwards compatibility, ran a local testnet to test the enr fork id and gossipsub topics switching logic.

Also, there's a potential incompatibility with the enr predicate that I have highlighted above.

AgeManning commented on beacon_node/eth2_libp2p/src/types/subnet.rs:
```diff
@@ -1,9 +1,23 @@
+#[derive(Debug, Clone, Copy, Serialize, PartialEq, Eq, Hash)]
+pub enum Subnet {
+    Attestation(SubnetId),
+    SyncCommittee(SubnetId),
```
Might be nice to add some doc comments here

AgeManning commented on beacon_node/network/src/router/mod.rs:
```diff
@@ -247,6 +247,8 @@ impl<T: BeaconChainTypes> Router<T> {
                     .on_attester_slashing_gossip(id, peer_id, attester_slashing);
             }
+            PubsubMessage::SignedContributionAndProof(_contribution_and_proof) => unimplemented!(),
+            PubsubMessage::SyncCommitteeSignature(_sync_committtee) => unimplemented!(),
```
these are actually unimplemented?

pawanjay176 replied:
Haven't implemented these yet. 
Need to build on top of https://github.com/sigp/lighthouse/pull/2321 . Have started some work on a local branch, but it's not complete yet.

AgeManning reviewed (commented):
This looks good to me. 

We need to do some extensive testing on this. The release tests seem to be failing, and a few minor comments.

pawanjay176 committed:
Address review comments

pawanjay176 committed:
Sync committee subnets

pawanjay176 committed:
Use SyncSubnetId for sync committee subnets

pawanjay176 committed:
Add worker tasks for sync committee gossip messages

pawanjay176 committed:
Use correct type level constants

pawanjay176 committed:
Decode altair gossip blocks

pawanjay176 committed:
Fix fork_digest in Status message

pawanjay176 committed:
Merge branch 'network-altair' into network-altair-gossip

pawanjay176 committed:
Add a SyncCommitteeSubscribe network event

michaelsproul commented on consensus/types/src/sync_subnet_id.rs:
```diff
@@ -33,6 +36,20 @@ impl SyncSubnetId {
+
+    /// Compute required subnets to subscribe to given the sync committee indices.
+    pub fn compute_subnets_for_sync_committee<T: EthSpec>(
+        sync_committee_indices: Vec<u64>,
```
I think this should be a slice

michaelsproul commented on consensus/types/src/sync_subnet_id.rs:
```diff
@@ -33,6 +36,20 @@ impl SyncSubnetId {
+        sync_committee_indices
+            .into_iter()
+            .map(|index| index.safe_div(subnet_size).map(Self::new))
+            .collect::<Result<_, _>>()
```
The spec [uses a set](https://github.com/ethereum/eth2.0-specs/blob/v1.1.0-alpha.7/specs/altair/validator.md#broadcast-sync-committee-message) for this return type, to account for the possibility of a validator being assigned multiple times to the same subnet. I implemented the same function for the VC, but will switch to using your implementation. I used `dedup` to get the set semantics (assuming `sync_committee_indices` is sorted).

```rust
    pub fn subnet_ids<E: EthSpec>(&self) -> Result<Vec<u64>, ArithError> {
        let mut subnets = self
            .validator_sync_committee_indices
            .iter()
            .map(|position| {
                position
                    .safe_div(E::SyncCommitteeSize::to_u64().safe_div(SYNC_COMMITTEE_SUBNET_COUNT)?)
            })
            .collect::<Result<Vec<_>, _>>()?;
        subnets.dedup();
        Ok(subnets)
    }
```

@realbigsean also has a version of this method that returns `HashMap<SyncSubnetId, Vec<usize>>`, but I'm happy to keep that separate (https://github.com/sigp/lighthouse/pull/2321#discussion_r647163764)

pawanjay176 replied:
We could also just return a `HashSet` from the function instead of a `Vec` right? Wouldn't have to make assumptions about `sync_committee_indices` being sorted then.

michaelsproul replied:
Yeah actually that's a lot safer

michaelsproul commented on consensus/types/src/chain_spec.rs:
```diff
@@ -437,6 +438,7 @@ impl ChainSpec {
             network_id: 1, // mainnet network id
             attestation_propagation_slot_range: 32,
             attestation_subnet_count: 64,
+            sync_committee_subnet_count: 4,
```
I think this should be deleted from `ChainSpec` because it's on the `EthSpec`.

We _also_ have a constant `SYNC_COMMITTEE_SUBNET_COUNT` in `consts.rs` which should probably be removed in favour of the `EthSpec` value (although that refactor will be a bit more involved).

pawanjay176 replied:
Yeah I added this because we had `attestation_subnet_count` in the `ChainSpec` and `SubnetBitfieldLength` in the `EthSpec` which is similar to this. I think we could have had it just in the `EthSpec` but wasn't so sure if there was a reason having it in both.

michaelsproul replied:
Oh right, interesting... I think lets clean up the sync one for now and consider coming back for the attestation subnet count later. Perhaps we could tolerate 2 representations (const and EthSpec), so long as we have a test somewhere that they're consistent?

michaelsproul reviewed (commented):
Just a few things I noticed while building VC stuff atop this

pawanjay176 committed:
Address review comments

pawanjay176 committed:
Fix sync committee subscriptions

pawanjay176 committed:
Add sync committee subscription test

pawanjay176 committed:
Merge branch 'network-altair' into network-altair-gossip

pawanjay176 committed:
Gossipsub message_id_fn changes

michaelsproul committed:
Expand and refactor fork functions

michaelsproul committed:
Import some relevant funcs from the VC

Based on "Expand and refactor fork functions"

realbigsean committed:
Updates related to merge with `unstable`

realbigsean committed:
Couple more `unstable` merge fixes

michaelsproul committed:
Simplify compute_subnets_for_sync_committee

realbigsean committed:
rust 1.53.0 updates (#2411)

## Issue Addressed

`make lint` failing on rust 1.53.0.

## Proposed Changes

1.53.0 updates

## Additional Info

I haven't figure out why yet, we were now hitting the recursion limit in a few crates. So I had to add `#![recursion_limit = "256"]` in a few places


Co-authored-by: realbigsean <seananderson33@gmail.com>
Co-authored-by: Michael Sproul <michael@sigmaprime.io>

michaelsproul commented:
I have a few commits on `altair-vc` that I wouldn't mind getting into this branch, in order to prevent future conflicts:

* This one related to fork names/epochs/etc on the ChainSpec: https://github.com/sigp/lighthouse/pull/2404/commits/834b5b9933be594414b24f86820f34065dee5c5d
* This one simplifying the arguments to `compute_subnets_for_sync_committeee`: https://github.com/sigp/lighthouse/pull/2404/commits/477f55423278b210b0834c41459534a8e744cc9c

@realbigsean has also updated his branch for the `SyncCommitteeSignature` to `SyncCommitteeMessage` rename, and included @divagant-martian's networking refactor from #2410. Pawan, if you could handle bubbling those changes up into this PR that would be great, then I'll update `altair-vc`.

pawanjay176 committed:
Merge branch 'network-altair' into network-altair-gossip

pawanjay176 committed:
Change libp2p dependency for message_id_fn

pawanjay176 committed:
appease clippy

pawanjay176 committed:
Add sync committee related metrics

pawanjay176 committed:
Fix subnet service test

pawanjay176 committed:
Add v2 protocols for OutboundRequests missed in the merge

michaelsproul committed:
Minify slashing protection interchange data (#2380)

## Issue Addressed

Closes #2354

## Proposed Changes

Add a `minify` method to `slashing_protection::Interchange` that keeps only the maximum-epoch attestation and maximum-slot block for each validator. Specifically, `minify` constructs "synthetic" attestations (with no `signing_root`) containing the maximum source epoch _and_ the maximum target epoch from the input. This is equivalent to the `minify_synth` algorithm that I've formally verified in this repository:

https://github.com/michaelsproul/slashing-proofs

## Additional Info

Includes the JSON loading optimisation from #2347

pawanjay176 committed:
Handle sync committee gossip errors

pawanjay176 committed:
Remove delay for updating fork

Commit:
Fix typo in advanced networking docs (#2412)

## Proposed Changes

Fixing a typo in the advanced networking docs which mentions ``--target-peer`` instead of the correct ``--target-peers`` flag

pawanjay176 committed:
Split rpc encode/decode into smaller functions

realbigsean committed:
fix the expected error in the sync contribution verification test at the sync period boundary

michaelsproul committed:
Merge remote-tracking branch 'origin/unstable' into altair

michaelsproul committed:
Fix issues with skip slots at fork boundaries

michaelsproul committed:
Fix clippy again

michaelsproul committed:
Merge remote-tracking branch 'origin/altair' into sync-committee-pools

realbigsean committed:
alpha.8 updates

pawanjay176 committed:
Pop sync committee tasks from beacon processor

pawanjay176 committed:
Fix some gossip scoring severity issue

pawanjay176 committed:
Merge branch 'sync-committee-pools' into network-altair-gossip

pawanjay176 committed:
Fix bug in peer_manager; Improve logging around subnet discoveries
